core/src/dsnp/graph_page.rs

```diff
@@ -100,32 +277,322 @@ impl GraphPage {
+		}
+
+		self.update_prids(prid_keys)?;
+
+		let DsnpUserPublicGraphChunk { compressed_public_graph } =
+			self.connections.clone().try_into()?;
+		let private_chunk = DsnpUserPrivateGraphChunk {
+			key_id,
+			prids: self.prids.clone(),
+			encrypted_compressed_private_graph: E::encrypt(&compressed_public_graph, key)?,
+		};
+
+		Ok(PageBlob {
+			page_id: self.page_id,
+			content_hash: self.content_hash,
+			payload: SchemaHandler::write_private_graph_chunk(&private_chunk)?,
+		})
+	}
 }
 
 impl Graph {
 	/// Create a new, empty Graph
-	pub fn new() -> Self {
-		Self { pages: HashMap::<PageId, GraphPage>::new() }
+	pub fn new(connection_type: ConnectionType) -> Self {
+		Self { connection_type, pages: PageMap::new() }
 	}
 
 	/// Getter for Pages in Graph
-	pub fn pages(&self) -> &HashMap<PageId, GraphPage> {
+	pub fn pages(&self) -> &PageMap {
 		&self.pages
 	}
 
 	/// Setter for Pages in Graph
-	pub fn set_pages(&mut self, pages: HashMap<PageId, GraphPage>) {
+	pub fn set_pages(&mut self, pages: PageMap) {
 		self.pages = pages;
 	}
 
+	/// Get next available PageId for this graph
+	pub fn get_next_available_page_id(&self) -> Option<PageId> {
+		const MAX_PAGE_ID: PageId = 16; // todo: move this to config
+		let existing_pages =
+			self.pages.iter().map(|(page_id, _)| *page_id).collect::<HashSet<PageId>>();
+		for pid in 0..MAX_PAGE_ID {
+			if !existing_pages.contains(&pid) {
+				return Some(pid)
+			}
+		}
+
+		None
+	}
+
+	/// Remove all pages from this graph
+	pub fn clear(&mut self) {
+		self.pages.clear();
+	}
+
+	/// Import bundle of pages as a Public Graph
+	pub fn import_public<E: EncryptionBehavior>(
+		&mut self,
+		ImportBundle::<E> { connection_type, pages, .. }: ImportBundle<E>,
+	) -> Result<()> {
+		if connection_type != self.connection_type {
+			return Err(Error::msg("Incorrect connection type for graph import"))
+		}
+
+		let mut page_map = HashMap::<PageId, GraphPage>::new();
+		for page in pages.iter() {
+			match GraphPage::try_from(page) {
+				Err(e) => return Err(e),
+				Ok(p) => {
+					page_map.insert(page.page_id, p);
+				},
+			};
+		}
+
+		self.set_pages(page_map);
+
+		Ok(())
+	}
+
+	/// Import bundle of pages as a Private Graph
+	pub fn import_private<E: EncryptionBehavior>(
+		&mut self,
+		ImportBundle::<E> { connection_type, pages, keys, .. }: ImportBundle<E>,
+	) -> Result<()> {
+		if connection_type != self.connection_type {
+			return Err(Error::msg("Incorrect connection type for graph import"))
+		}
+
+		let mut page_map = PageMap::new();
+		for page in pages.iter() {
+			match GraphPage::try_from((page, &keys)) {
+				Err(e) => return Err(e),
+				Ok(p) => {
+					page_map.insert(page.page_id, p);
+				},
+			};
+		}
+
+		self.set_pages(page_map);
+
+		Ok(())
+	}
+
+	/// Import bundle of pages as a Private Graph, but without decrypting the encrypted portion
+	/// Useful for importing a foreign user's graph for inspection of PRIds
+	pub fn import_opaque<E: EncryptionBehavior>(
```

Comment:
I'm thinking about a PRIManager or something like that to facilitate the PRI calculation